embedded-tests/src/test/java/org/apache/druid/testing/embedded/indexing/NestedDataFormatsTest.java

```diff
@@ -0,0 +1,146 @@
+
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Embedded tests for nested data, ingested in different {@link NestedCommonFormatColumnFormatSpec}.
+ */
+public class NestedDataFormatsTest extends EmbeddedClusterTestBase
+{
+  private final EmbeddedBroker broker = new EmbeddedBroker();
+  private final EmbeddedOverlord overlord = new EmbeddedOverlord();
+  private final EmbeddedCoordinator coordinator = new EmbeddedCoordinator();
+
+  private final String defaultFormat = "koala_default";
+
+  @Override
+  protected EmbeddedDruidCluster createCluster()
+  {
+    return EmbeddedDruidCluster.withEmbeddedDerbyAndZookeeper()
+                               .useLatchableEmitter()
+                               .useDefaultTimeoutForLatchableEmitter(60)
+                               .addServer(overlord)
+                               .addServer(coordinator)
+                               .addServer(new EmbeddedIndexer())
+                               .addServer(new EmbeddedHistorical())
+                               .addServer(broker)
+                               .addServer(new EmbeddedRouter());
+  }
+
+  @BeforeAll
+  protected void ingestWithDefaultFormat() throws Exception
+  {
+    final TaskBuilder.IndexParallel indexTask =
+        TaskBuilder.ofTypeIndexParallel()
+                   .dataSource(defaultFormat)
+                   .timestampColumn("timestamp")
+                   .jsonInputFormat()
+                   .inputSource(Resources.HttpData.kttm1Day())
+                   .schemaDiscovery();
+
+    final String taskId = EmbeddedClusterApis.newTaskId(defaultFormat);
+    cluster.callApi().runTask(indexTask.withId(taskId), overlord);
+    cluster.callApi().waitForAllSegmentsToBeAvailable(defaultFormat, coordinator, broker);
+  }
+
+  @Test
+  public void test_objectStorageEncoding()
+  {
+    // Ingest kttm data with skipping smile raw json format, comparing diff with defaultFormat
+    NestedCommonFormatColumnFormatSpec spec =
+        NestedCommonFormatColumnFormatSpec.builder().setObjectStorageEncoding(ObjectStorageEncoding.NONE).build();
+    final TaskBuilder.IndexParallel indexTask =
+        TaskBuilder.ofTypeIndexParallel()
+                   .dataSource(dataSource)
+                   .timestampColumn("timestamp")
+                   .jsonInputFormat()
+                   .inputSource(Resources.HttpData.kttm1Day())
+                   .schemaDiscovery()
+                   .tuningConfig(t -> t.withIndexSpec(IndexSpec.builder().withAutoColumnFormatSpec(spec).build()));
+    final String taskId = EmbeddedClusterApis.newTaskId(dataSource);
+    cluster.callApi().runTask(indexTask.withId(taskId), overlord);
+    cluster.callApi().waitForAllSegmentsToBeAvailable(dataSource, coordinator, broker);
+
+    final String resultAsJson =
+        cluster.callApi().onAnyBroker(b -> b.submitSqlQuery(ClientSqlQuery.simple("select * from sys.segments")));
```

Comment:
Why not use a 
```sql
SELECT COUNT(*), SUM(num_rows) FROM sys.segments WHERE datasource = '%s'
```

You can fire a separate query for each datasource. I feel the code would be cleaner that way.

Reply:
You can also just do `cluster.runSql(sqlString)` to get back a CSV response.